docs/rfcs/0006-top-sites-feature.md

```diff
@@ -0,0 +1,125 @@
+---
+layout: page
+title: Introduce a Top Sites Feature
+permalink: /rfc/0006-top-sites-feature
+---
+
+* Start date: 2020-07-30
+* RFC PR: [#7931](https://github.com/mozilla-mobile/android-components/pull/7931)
+
+## Summary
+
+Create a TopSitesFeature that abstracts out the co-ordination between the multiple data storages.
+
+## Motivation
+
+A large part of the top sites implementation is spread across A-C and Fenix. There is a lot of co-ordination between the History API and the current Top Sites API to manage the final state of what sites should be displayed that can exist as a single implementation.
+
+## Guide-level explanation
+
+To begin, we should rename the `TopSitesStorage`  to `PinnedSitesStorage` and create a `TopSitesFeature` that would orchestrate the different storages and present them to the app.
+
+We can then follow the architecture of existing features by using the presenter/interactor/view model, and introduce a `TopSitesStorage` which abstract out the complexity of the multiple data sources from `PinnedSitesStorage` and `PlacesHistoryStorage`.
+
+To allow the ability for adding a top site from different parts of the app (e.g. context menus), we introduce the `TopSitesUseCases` that acts on the storage directly. If the `TopSitesFeature` is started, it will then be notified by the `Observer` to update the UI.
+
+```kotlin
+
+/**
+ * Implemented by the application for displaying onto the UI.
+ */
+interface TopSitesView {
+  /**
+   * Update the UI.
+   */
+  fun displaySites(sites: List<TopSite>)
+
+  interface Listener {
+    /**
+     * Invoked by the UI for us to add to our storage.
+     */
+    fun onAddTopSite(topSite: TopSite)
+
+    /**
+     * Invoked by the UI for us to remove from our storage.
+     */
+    fun onRemoveTopSite(topSite: TopSite)
+  }
+}
+
+/**
+ * Abstraction layer above the multiple storages.
+ */
+interface TopSitesStorage {
+
+  /**
+   * Return a unified list of top sites based on the given number of sites desired.
+   * If `includeFrecent` is true, fill in any missing top sites with frecent top site results.
+   */
+  fun getTopSites(totalNumberOfSites: Int, includeFrecent: Boolean): List<TopSite>
+
+  interface Observer {
+    /**
+     * Invoked when changes are made to the storage.
+     */
+    fun onStorageUpdated()
+  }
+}
+
+// Already exists in browser-storage-sync.
+interface PlacesHistoryStorage {
+  fun getTopFrecentSites(num: Int)
+}
+
+class DefaultTopSitesStorage(
+  val pinnedSitesStorage: PinnedSitesStorage,
+  val historyStorage: HistoryStorage? = null
+) : TopSitesStorage {
+
+  /**
+   * Merge data sources here, return a single list of top sites.
+   */
+  override fun getTopSites(numFrecent): List<TopSite>
+}
+
+/**
+ * Use cases can be used for adding a top from different places like a context menu.
+ */
+class TopSitesUseCases(topSitesStorage: TopSitesStorage) {
```

Comment:
Again, `PinnedSitesUseCases`, since you're only modifying a single storage. (not frecent sites, not default sites).

Reply:
We left the use cases as top sites as well based on the known problems listed in the RFC: how do we delete a frecent site?

@gabrielluong had made a good point that we don't want to alter the frecent data from history since we are just a consumer of the data and our use case (top sites) shouldn't alter the frecent data that might be valid in a different context.

Reply:
Currently, "deleting frecent sites" will look like storing an exclusion list of sites that you'd filter out if they come up as results from the frecent storage. This could live in PlacesHistoryStorage right on top of a-s, or even in a-s itself, and provide a nice "ignore frecent site" api. cc @linacambridge

Reply:
At first, we were thinking of having a-c store the exclusion list, since it's already storing the pinned list.

But thinking more about it, I wonder if we want to push down pinned _and_ ignored sites into a-s. IOW, we end up with a new tri-state integer column in `moz_places`: -1 for "never show in top sites", 0 for "no preference, show if frecent enough", and 1 for "always show".

Then the query to fetch top sites looks something like:

```sql
-- First, select all sites that the user pinned.
SELECT url, title, pin, frecency FROM moz_places
WHERE pin = 1
UNION ALL
-- Fill the rest with top frecent sites.
SELECT url, title, pin, frecency FROM moz_places 
WHERE pin != -1 AND
      (SUBSTR(url, 1, 6) == 'https:' OR SUBSTR(url, 1, 5) == 'http:') AND
      (last_visit_date_local + last_visit_date_remote) != 0 AND
      NOT hidden
-- User-pinned sites first (1 > 0), then most frecent, up to the limit.
ORDER BY pin, frecency DESC
LIMIT :num
```

WDYT? I think if we're going to track ignored sites in a-s, we might as well go all the way and track pinned ones, too. Having an `ignoreFrecentSite` (`ignoreTopSite`?) method without a corresponding `pinTopSite` feels asymmetrical.

Reply:
My immediate thoughts were that this would be a very interest approach because we can essentially remove `PinnedSitesStorage` from A-C and just focus on the sites provided by a single storage `PlacesHistoryStorage`. 

Thinking from an outside A-C consumer perspective, I am thinking the `TopSitesFeature` would create a bit more overhead if I wanted to use my own roomDB, but would now need to tie in with the `PlacesHistoryStorage` (A-S storage - no idea if this means setting up their own A-S).

Reply:
We are currently working under the assumption that we have frecent top sites in Fenix by the end of August but ideally as early as mid August. I think we need to create a couple of scenarios / steps / chunks of work and see what is most feasible given the timeline, and plan any remaining follow up work as necessary. Of course, before doing that I think settling on a solution comes first.

Reply:
Questions I would have if we considered this approach of migrating to keep all sites in `PlacesHistoryStorage`:

- How hard would it be to migrate existing top sites from `TopSitesStorage` to `PlacesHistoryStorage`? Where do we think default top sites provided by the application should go - maybe this goes back to the defaultSites list that is provided by the application, but I think it would need its own exclusion list in the event that an user wants to exclude seeing a particular default site.
- Let's examine the current "Remove from history" context menu action in Desktop / Fennec and re-evaluate if the exclusion list is still the right solution here vs doing the same.